compiler/rustc_codegen_ssa/src/mir/mod.rs

```diff
@@ -171,6 +172,32 @@ pub fn codegen_mir<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
     let fn_abi = cx.fn_abi_of_instance(instance, ty::List::empty());
     debug!("fn_abi: {:?}", fn_abi);
 
+    if cx.tcx().has_attr(instance.def.def_id(), rustc_span::sym::naked) {
+        let cached_llbbs = IndexVec::new();
+
+        let fx: FunctionCx<'_, '_, Bx> = FunctionCx {
+            instance,
+            mir,
+            llfn,
+            fn_abi,
+            cx,
+            personality_slot: None,
+            cached_llbbs,
+            unreachable_block: None,
+            terminate_block: None,
+            cleanup_kinds: None,
+            landing_pads: IndexVec::from_elem(None, &mir.basic_blocks),
+            funclets: IndexVec::from_fn_n(|_| None, mir.basic_blocks.len()),
+            locals: locals::Locals::empty(),
+            debug_context: None,
+            per_local_var_debug_info: None,
+            caller_location: None,
+        };
+
+        fx.codegen_naked_asm(instance);
+        return;
+    }
```

Comment:
this branch is weird of course. We could add an extra `NakedAsm` variant to `MonoItem` to make this a bit nicer maybe?

Reply:
This `cx.tcx().has_attr(instance.def.def_id(), rustc_span::sym::naked)` panics, apparently because the `naked` attribute is `EncodeCrossCrate::No`. I'm not sure why that is relevant here. 

```
12:     0x7f173a21e691 - <rustc_middle[ac0ad9b6bcaf7b38]::ty::context::TyCtxt>::get_attrs::<rustc_span[2f17d50bcbc9f57a]::def_id::DefId>
13:     0x7f173a3ac676 - rustc_codegen_ssa[4c7413a1cf9b9a7e]::mir::codegen_mir::<rustc_codegen_llvm[12d34c80ef723660]::builder::Builder>
14:     0x7f173a257d28 - rustc_codegen_ssa[4c7413a1cf9b9a7e]::base::codegen_instance::<rustc_codegen_llvm[12d34c80ef723660]::builder::Builder>
15:     0x7f173a2d8dd9 - <rustc_middle[ac0ad9b6bcaf7b38]::mir::mono::MonoItem as rustc_codegen_ssa[4c7413a1cf9b9a7e]::mono_item::MonoItemExt>::define::<rustc_codegen_llvm[12d34c80ef723660]::builder::Builder>
```

so, is a fundamentally different approach needed here, or would e.g. a check on

```rust
let attrs = tcx.hir().attrs(tcx.local_def_id_to_hir_id(did));
```

be OK and circumvent the assert?